### PR TITLE
Enable RequestMoney in Workspace chats

### PR DESCRIPTION
**Pull request by luacmartins**
### Details
Allows users to request money from a Workspace chat.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/270581

### Tests
Pre-requisite: 2 accounts

1. From the admin account, create a workspace `+ > New workspace`
2. Invite the employee account to the workspace, `Settings > Workspaces > Manage member > Invite`
3. Log into the employee account
4. Find the workspace chat in the LHN
5. Tap `+ > Request money`
6. Verify that you can request money from the workspace
7. Refresh the page and verify that the request is still there
8. Turn off your internet connect
9. Repeat steps 5-6 and verify that the request goes through but it's now greyed out
10. Turn on internet connection
11. Verify that the request is no longer greyed out
12. Refresh the page and verify that nothing changes
13. Repeat steps 3-12, but start the Request money from from the global create (gloating green `+` button) instead of the chat `+` button and select the Workspace chat as the participant.

- [ ] Verify that no errors appear in the JS console

### Offline tests
Steps above.

### QA Steps
Test steps above

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/22219519/235985005-bbf39d85-9852-475d-bbb9-2a0969c39e79.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/22219519/235985027-dccb66e9-1356-4747-8a78-9ba1816a4a4f.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/22219519/235985046-e1dae443-e9f9-436f-8f05-b51512bd12de.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/22219519/235985058-648f5741-2bd8-4a3f-b5b9-d187a00e4bbc.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/22219519/235985076-ba27d2c3-5bc0-4443-b2c1-8939aa6b5dbe.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/22219519/235985094-2bcfd609-7f6d-4fbd-883c-0ff9140484db.mov

</details>

### CODE REVIEW

**Comment by luacmartins** [comment]
I think this is ready for review1

**Review by mountiny — DISMISSED** [review]
Code looks good to me, gonna have to test, I htink this is not ready for testing yet @parasharrajat

**Review by danieldoglas — DISMISSED** [review]
Code LGTM as well

**Comment by luacmartins** [comment]
@parasharrajat can you prioritize reviewing this PR today? This is of high value and is blocking other issues in our roadmap pipeline with an external commit of May 17

**Comment by s77rt** [comment]
@luacmartins 

> Find the workspace chat in the LHN


I can't find it :sweat_smile: Am I missing something here?

**Comment by luacmartins** [comment]
Have we added your test account part of the policyExpenseChat beta? If not, can you please ask in Slack if someone can add you to it? My flight is about to take off and I'll lose connection

**Comment by s77rt** [comment]
Getting invalid email error message, not sure if it's related to the account being newly added to the beta.

https://user-images.githubusercontent.com/16493223/236296529-bc22860c-9d19-4524-881b-cdee11a2bdf7.mp4

**Comment by mountiny** [comment]
@s77rt @parasharrajat I have synced this up with main so you can test with no issue now! Thank you very much! 

Also started the internal build if thats easier to test https://github.com/Expensify/App/actions/runs/4888345058

**Review by s77rt — COMMENTED** [review]
**Bugs**
1. The IOU preview shows wrong info (looks we are setting incorrect optimistic data)
    1. Showing "paid you" instead of "owes you"
    2. The amount is shown as is without conversion
2. After page refresh the IOU preview is hidden


https://user-images.githubusercontent.com/16493223/236357950-6a97a0a2-b4a2-4b7e-8bbb-240334b54a4e.mp4

**Comment by luacmartins** [comment]
@s77rt [this error](https://github.com/Expensify/App/pull/18215#issuecomment-1535228956) is because the API changes weren't in production yet. I checked the logs and that request went to the production server. The changes have since been deployed to production. Can you try it again?

**Comment by s77rt** [comment]
@luacmartins Yes, I was able to get this working. I reported some bugs above.

and we have some conflicts :grin:

**Comment by luacmartins** [comment]
@s77rt 

>The IOU preview shows wrong info (looks we are setting incorrect optimistic data)
>Showing "paid you" instead of "owes you"
>The amount is shown as is without conversion

I wasn't able to reproduce this. Does this always happen for you?


>After page refresh the IOU preview is hidden

We are changing the logic to always display the preview, so I'd say this is not an issue. Additionally, I did find a bug but it is in the API so I'll address that in a different PR.

**Comment by s77rt** [comment]
@luacmartins 

> I wasn't able to reproduce this. Does this always happen for you?

Yes, try request money offline. I requested €5, the preview is showing $5 (the conversion is not right) and having the "paid you" comment.


https://user-images.githubusercontent.com/16493223/236436283-711243e5-1261-48f8-ba0c-fe84d8bda280.mp4

**Review by mountiny — COMMENTED** [review]
@luacmartins Tested and it works fine if requested from workspace chat, but if I request it from the global create and select the workspace, it does not land in the correct chat, created a new chat in fact which shows error. Is this expected, should we cover this case of creating a request from global create too?

https://user-images.githubusercontent.com/36083550/236479523-66c2bb8d-4f73-496b-9608-2ac367989e64.mp4

**Comment by mountiny** [comment]
@s77rt can you reproduce what I saw above?

**Comment by mountiny** [comment]
Also agree with what @luacmartins written above

**Comment by s77rt** [comment]
@mountiny Yes, same.

**Comment by s77rt** [comment]
May be unrelated but sometimes I get the delete option even for already deleted transactions
Screenshot from 2023-05-05 15-56-37

**Comment by s77rt** [comment]
^ Maybe from the backend? I'm not getting the "delete" report actions
Screenshot from 2023-05-05 16-11-32

**Review by Julesssss — COMMENTED** [review]
it isn't clear to me why we can remove `participants` and how we will build this data elsewhere. Please let me know if I missed the obvious answer. 

I'm mostly asking because I believe we're going to need to start using `participants` to build the IOUPreview icons for a bill split

**Comment by luacmartins** [comment]
@mountiny @s77rt addressed the global create issue

**Comment by luacmartins** [comment]
@Julesssss I only removed `participants` from `buildOptimisticIOUReportAction` because we removed the participants from the `getIOUReportActionMessage` message templates, so this param was unused now. We can always reintroduce it if there's a need for it again.

**Comment by luacmartins** [comment]
Ready for another review!

**Comment by s77rt** [comment]
@luacmartins Can you confirm if the missing "delete" actions is backend related (https://github.com/Expensify/App/pull/18215#issuecomment-1536406907)

**Comment by luacmartins** [comment]
@s77rt yes, that seems to be an API issue. I can reproduce when I log out and log back in. All delete actions are missing from the server response. I don't think that's an issue with this PR. I'll investigate the bug on the API layer.

**Comment by s77rt** [comment]
@luacmartins Thanks for the confirmation. Looks we are almost done here. Will take another look then approve asap.

**Comment by luacmartins** [comment]
Found the bug on our server, it's only affecting requests on Workspace chats, regular IOUs should return those deleted messages fine. I'll put up a PR to fix it.

**Comment by mountiny** [comment]
Getting some weird behaviour but trying to confirm now its also on auth main
image

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:59]:
```suggestion
    const isPolicyExpenseChat = report.isPolicyExpenseChat || report.isOwnPolicyExpenseChat;
```

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:64]: This is super confusing, why is it participant?
```suggestion
        chatReport = chatReports[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`];
```

**Review by mountiny — COMMENTED** [review]
one question

**Review by mountiny — CHANGES_REQUESTED** [review]
I think we need to update the Auth change to make sure the expense transactions are always added as negative values. Thats how it's aligned with previous expense logic and expenes saved. we should nto deploy so people's datas are not messed up

**Comment by mountiny** [comment]
Made a PR here https://github.com/Expensify/Auth/pull/7860

File: src/libs/actions/IOU.js
Comment (s77rt) [reply src/libs/actions/IOU.js:59]: This won't work if requesting from global menu (no report).

Comment (luacmartins) [reply src/libs/actions/IOU.js:59]: Correct, we need to use the selected participant instead of the report because as @s77rt pointed out, there's no report when we use the global create.

File: src/libs/actions/IOU.js
Comment (luacmartins) [reply src/libs/actions/IOU.js:64]: Because the report is null if we initiate the flow from the global create. However, the participant always stores the reportID if it exists.

Comment (luacmartins) [reply src/libs/actions/IOU.js:64]: I added a comment.

**Review by mountiny — DISMISSED** [review]
Ok cool, thank you! The Auth PR will be deployed probably on Wednesday

**Review by Julesssss — DISMISSED** [review]
Reviewing the latest changes. I was going to suggest improvements to the comment, but it honestly doesn't seem worth spending time on. Let's merge once the Auth PR is deployed

**Comment by JmillsExpensify** [comment]
With Auth merged, can we take this one off hold now?

**Comment by s77rt** [comment]
Someone please tag me once this is no longer on hold (as we don't get notifications for title change).

**Review by s77rt — DISMISSED** [review]
LGTM! :rocket:

**Comment by s77rt** [comment]
@luacmartins Thanks! I have approved the PR. Let me know if there anything else is needed from my side.

**Comment by JmillsExpensify** [comment]
Thanks all!

**Comment by luacmartins** [comment]
@s77rt @Julesssss @mountiny Had to resolve some style conflicts coming from the Prettier change we just introduced. Can I get another approval?

**Review by Julesssss — APPROVED** [review]
LGTM

**Comment by Julesssss** [comment]
The Auth PR has been merged so i think we're just awaiting deployment?

**Comment by luacmartins** [comment]
Yea, just waiting deploy which should happen soon!

**Review by s77rt — APPROVED** [review]

**Comment by s77rt** [comment]
@luacmartins Requesting money using admin account shows both the workspace and the members as participants. Seems like a bug.


https://github.com/Expensify/App/assets/16493223/b63e5659-2e99-452e-b4be-2355a6b51a5f

**Review by s77rt — CHANGES_REQUESTED** [review]
Address bug https://github.com/Expensify/App/pull/18215#issuecomment-1541909985

**Review by mountiny — APPROVED** [review]
Auth deployed

**Comment by mountiny** [comment]
I think this is an edge case and we can address it in a follow up

File: src/libs/ReportUtils.js
Comment (situchan) [inline src/libs/ReportUtils.js:1154]: This change needs reverted.
Right now app crashes on `main` because of this line which was added in https://github.com/Expensify/App/pull/18656
https://github.com/Expensify/App/blob/1b1e562a33c01dbbc2ca81c8994342f95b5aa481/src/libs/ReportUtils.js#L1181

Comment (Beamanator) [reply src/libs/ReportUtils.js:1154]: Let's discuss a resolution in slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1683719211068359

File: src/libs/ReportUtils.js
Comment (aldo-expensify) [inline src/libs/ReportUtils.js:1082]: The `state` and the `stateNum` are not the same, I think they should be the same right?
I also think that `state` is not really used, maybe we should remove it.

This is also inconsistent here:

image

copy/paste?

Comment (mountiny) [reply src/libs/ReportUtils.js:1082]: hmm I dont think they should be same/identical, right? @luacmartins

**Pull request merged**
